### DISCUSSION TITLE
Copy-to-clipboard Macro and the Hardcoded Button Caption

### DISCUSSION THREAD

**Discussion by kookma** [TiddlyWiki/TiddlyWiki5#6774] [Ideas]
In https://tiddlywiki.com/prerelease/#%24%3A%2Fcore%2Fmacros%2Fcopy-to-clipboard the handy macro `copy-to-clipboard` has a hardcoded caption. see

```
<$button class=<<__class__>> style=<<__style__>> message="tm-copy-to-clipboard" param=<<__src__>> tooltip={{$:/language/Buttons/CopyToClipboard/Hint}}>
{{$:/core/images/copy-clipboard}}
&#32;
<$text text={{$:/language/Buttons/CopyToClipboard/Caption}}/>
</$button>
```

and the caption: `<$text text={{$:/language/Buttons/CopyToClipboard/Caption}}/>`

But there are many use cases where, only the icon is enough!

So my suggestion is

1. lets pass the button caption while it can have a default value, OR
2. lets caption appears based on some global setting used for other UI buttons

What do you think?

**Comment by pmario** [comment]
I think it needs to be a "per button" setting, otherwise it wouldn't be backwards compatible. 

There are 2 possibilities

1.  Allow the user to define the caption text
2.  Switch on / off the existing translated caption text. 

I personally would suggest to use option 2, since it works with translated labels. User defined labels are 1 language only. Due to the meaning of the icon itself, there shouldn't be different user defined texts. 

What do you think? .. The implementation is relatively straight forward. 

POC

```
\define getCaption(showCaption:"yes")
<$list filter="[<__showCaption__>match[yes]]" variable=ignore>
&#32;
<$text text={{$:/language/Buttons/CopyToClipboard/Caption}}/>
</$list>
\end
\define copy-to-clipboard(src,class:"tc-btn-invisible",style,showCaption)
\whitespace trim
<$button class=<<__class__>> style=<<__style__>> message="tm-copy-to-clipboard" param=<<__src__>> tooltip={{$:/language/Buttons/CopyToClipboard/Hint}}>
{{$:/core/images/copy-clipboard}}
<$macrocall $name="getCaption" showCaption=<<__showCaption__>>/>
</$button>
\end

\define copy-to-clipboard-above-right(src,class:"tc-btn-invisible",style,showCaption)
\whitespace trim
<div style="position: relative;">
<div style="position: absolute; bottom: 0; right: 0;">
<$macrocall $name="copy-to-clipboard" src=<<__src__>> class=<<__class__>> style=<<__style__>> showCaption=<<__showCaption__>>/>
</div>
</div>
\end
```

macrocall:

```
<<copy-to-clipboard """
test no caption
""" showCaption:no>>

<<copy-to-clipboard """
test with caption
""" >>

```

**Comment by Jermolene** [comment]
The copy-to-clipboard macro is pretty simple. Rather than risking making it more complex by adding more parameters and options, I'd be tempted to explore your use cases via alternate versions of the macro that are completely separate. For example, `<<copy-to-clipboard-small>>` for a version that doesn't use the text caption.

**Comment by pmario** [comment]
probably related:  proposal to improve copy-to-clipboard #4463

**Comment by pmario** [comment]
I did think about it again. The problem is I do want to have a maximum flexible drop-in-replacement for the tw5-com copy-to-clipboard macro, without duplicating any code. 

I did take the code my proposal from: **proposal to improve copy-to-clipboard** #4463  and added 2 new macros, which have a simple interface and copy-to-clipboard has the flexibel interface. 

It can

- use language specific caption by default, but can change the caption
- use default icon, but can change the icon
- add a copy-to-clipboard button to  `code like this` 

![image](https://user-images.githubusercontent.com/374655/206221710-c5f4e908-e78a-40bf-b07d-c3b0704376d1.png)

---------

On the other hand the existing doc-macros need an overhaul anyway. .. They still use `$text-substitution$` which makes them brittle.
